src/compiler/checker.ts

```diff
@@ -27359,78 +27587,197 @@ namespace ts {
+                // The "return" and "next" types of `Iterable` and `IterableIterator` are defined by the
+                // iteration types of their `[Symbol.iterator]()` method. The same is true for their async cousins.
+                // While we define these as `any` and `undefined` in our libs by default, a custom lib *could* use
+                // different definitions.
+                const globalIterationTypes =
+                    getIterationTypesOfIterableCached(globalType, resolver) ||
+                    getIterationTypesOfIterableSlow(globalType, resolver, /*errorNode*/ undefined);
+                const { returnType, nextType } = globalIterationTypes === noIterationTypes ? defaultIterationTypes : globalIterationTypes;
+                return (type as IterableOrIteratorType)[resolver.iterableCacheKey] = createIterationTypes(yieldType, returnType, nextType);
+            }
+
+            // As an optimization, if the type is an instantiation of the following global type, then
+            // just grab its related type arguments:
+            // - `Generator<T, TReturn, TNext>` or `AsyncGenerator<T, TReturn, TNext>`
+            if (isReferenceToType(type, resolver.getGlobalGeneratorType(/*reportErrors*/ false))) {
+                const [yieldType, returnType, nextType] = (type as GenericType).typeArguments!;
+                return (type as IterableOrIteratorType)[resolver.iterableCacheKey] = createIterationTypes(yieldType, returnType, nextType);
+            }
+        }
+
+        /**
+         * Gets the *yield*, *return*, and *next* types of an `Iterable`-like or `AsyncIterable`-like
+         * type from its members.
+         *
+         * If we successfully found the *yield*, *return*, and *next* types, an `IterationTypes`
+         * record is returned. Otherwise, `noIterationTypes` is returned.
+         *
+         * NOTE: You probably don't want to call this directly and should be calling
+         * `getIterationTypesOfIterable` instead.
+         */
+        function getIterationTypesOfIterableSlow(type: Type, resolver: IterationTypesResolver, errorNode: Node | undefined) {
+            const method = getPropertyOfType(type, getPropertyNameForKnownSymbolName(resolver.iteratorSymbolName));
+            const methodType = method && !(method.flags & SymbolFlags.Optional) ? getTypeOfSymbol(method) : undefined;
+            if (isTypeAny(methodType)) {
+                return (type as IterableOrIteratorType)[resolver.iterableCacheKey] = anyIterationTypes;
+            }
+
+            const signatures = methodType ? getSignaturesOfType(methodType, SignatureKind.Call) : undefined;
+            if (!some(signatures)) {
+                return (type as IterableOrIteratorType)[resolver.iterableCacheKey] = noIterationTypes;
+            }
+
+            const iteratorType = getUnionType(map(signatures, getReturnTypeOfSignature), UnionReduction.Subtype);

```

Comment:
`getReturnTypeOfSignature`, hmmm. What happens were I to define an iterable iterator like
```ts
class F<T> {
   constructor(private value: T) {  }
   next() {
    return { done: true, value: this.value }
  }
  [Symbol.iterator]<U = string>(cast: (x: T) => U = String as any) { return new F(cast(this.value)) }
}
```
is the parameter on the `Symbol.iterator` defaulted correctly?

Reply:
We used `getReturnTypeOfSignature` in the original logic as well.

```ts
class F<T> {
	constructor(private value: T) { }
	next() {
		return { done: true, value: this.value }
	}
	[Symbol.iterator]<U = string>(cast: (x: T) => U = String as any) {
		return new F(cast(this.value))
	}
}

function * g() {
	const x1 = yield* new F(1);
}

for (const x2 of new F(1)) {
}
```

Both `x1` and `x2` have the type `string` because your `next` method isn't typed. What are you expecting to happen? Should I be doing something other than calling `getReturnTypeOfSignature` here?

Reply:
> because your next method isn't typed

? It should be `{done: true, value: T}` - You mean because `cast` isn't passed, right? Because that's what I hope would happen when it's not 😄

Reply:
No, the return type of `next` is widened to `{ done: boolean, value: T }`.